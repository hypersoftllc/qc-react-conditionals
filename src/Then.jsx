import React from 'react'


function Then(props) {
  let children = props.children,
      typeOf = typeof children

  return typeOf === 'number' || typeOf === 'string' ? children : children || null
}

export default Then
