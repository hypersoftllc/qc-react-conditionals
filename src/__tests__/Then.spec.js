/* eslint-env jest, node */
/* eslint max-len: "off" */
/* eslint padded-blocks: "off" */

import React from 'react'
import ErrorBoundary from 'react-error-boundary'
import TestRenderer from 'react-test-renderer'

import Then from '../Then'


describe('Then.jsx', function () {

  describe('Then', function () {

    it('should render with no props and no children', function () {
      let comp

      comp = TestRenderer.create(
        <Then/>
      )

      expect(comp.toJSON()).toMatchSnapshot()
    })

    it('should render a single `false` child as `null`', function () {
      let comp


      comp = TestRenderer.create(
        <Then>{ false }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{ true && false }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ false }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ true && false }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()
    })

    it('should render a single `true` child as `null`', function () {
      let comp


      comp = TestRenderer.create(
        <Then>{ true }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{ true || false }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ true }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ true || false }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()
    })

    it('should render a single `Date` child as `.toValue().toString()`', function () {
      let comp, date

      date = Date.UTC(2000, 0, 1)

      comp = TestRenderer.create(
        <Then>{ date }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ date }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()
    })

    it('should render a single non-zero number child as its string value', function () {
      let comp


      comp = TestRenderer.create(
        <Then>{-Infinity}</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{-Number.MAX_VALUE}</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{-Number.MAX_SAFE_INTEGER}</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{-42}</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{-1}</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{1}</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{42}</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{Number.MAX_SAFE_INTEGER}</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{Number.MAX_VALUE}</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{Infinity}</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ 42 }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()
    })

    it('should render a single zero child as `0`', function () {
      let comp


      comp = TestRenderer.create(
        <Then>{ 0 }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{ 1 - 1 }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ 0 }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()
    })

    it('should render a single `NaN` child as "NaN"', function () {
      let comp


      comp = TestRenderer.create(
        <Then>{ NaN }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{ parseInt('asdf', 10) }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ NaN }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()
    })

    it('should render a single non-empty string child as its string value', function () {
      let comp


      comp = TestRenderer.create(
        <Then>0</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{ '0' }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>non-empty</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{ 'non-empty' }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{ `non-empty` }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{ `non` + '-' + "empty" }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()
    })

    it('should render a single empty string child as ""', function () {
      let comp


      comp = TestRenderer.create(
        <Then>{ '' }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{ '  '.trim() }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ '' }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()
    })

    it('should render a single `undefined` child as `null`', function () {
      let comp


      comp = TestRenderer.create(
        <Then>{ undefined }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <Then>{ {}.foo }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ undefined }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()
    })

    it('should throw with a single `Error` child', function () {
      let comp, consoleErrorOriginal, onError

      consoleErrorOriginal = console.error
      try {
        // Prevent error message in console during test runs
        console.error = () => {}
        onError = jest.fn()
        comp = TestRenderer.create(
          <ErrorBoundary onError={onError}>
            <Then>{ [Error('Oh oh!')] }</Then>
          </ErrorBoundary>
        )

        expect(onError).toHaveBeenCalled()
        expect(comp.toJSON()).toMatchSnapshot()
      }
      finally {
        console.error = consoleErrorOriginal
      }
    })

    it('should render an empty array child as if there were no children', function () {
      let comp


      comp = TestRenderer.create(
        <Then>{ [] }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ [] }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      expect(comp.toJSON()).toEqual(TestRenderer.create(<div>{ [] }</div>).toJSON())
    })

    it('should render an array child as expected based on type', function () {
      let comp


      comp = TestRenderer.create(
        <Then>{ ['Hello!', ' ', 'Today I am ', 42, ' years old!'] }</Then>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      comp = TestRenderer.create(
        <div><Then>{ ['Hello!', ' ', 'Today I am ', 42, ' years old!'] }</Then></div>
      )

      expect(comp.toJSON()).toMatchSnapshot()


      expect(comp.toJSON()).toEqual(TestRenderer.create(<div>{ ['Hello!', ' ', 'Today I am ', 42, ' years old!'] }</div>).toJSON())
    })

  })

})
