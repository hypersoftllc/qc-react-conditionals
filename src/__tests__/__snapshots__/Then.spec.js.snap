// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Then.jsx Then should render a single \`Date\` child as \`.toValue().toString()\` 1`] = `"946684800000"`;

exports[`Then.jsx Then should render a single \`Date\` child as \`.toValue().toString()\` 2`] = `
<div>
  946684800000
</div>
`;

exports[`Then.jsx Then should render a single \`NaN\` child as "NaN" 1`] = `"NaN"`;

exports[`Then.jsx Then should render a single \`NaN\` child as "NaN" 2`] = `"NaN"`;

exports[`Then.jsx Then should render a single \`NaN\` child as "NaN" 3`] = `
<div>
  NaN
</div>
`;

exports[`Then.jsx Then should render a single \`false\` child as \`null\` 1`] = `null`;

exports[`Then.jsx Then should render a single \`false\` child as \`null\` 2`] = `null`;

exports[`Then.jsx Then should render a single \`false\` child as \`null\` 3`] = `<div />`;

exports[`Then.jsx Then should render a single \`false\` child as \`null\` 4`] = `<div />`;

exports[`Then.jsx Then should render a single \`true\` child as \`null\` 1`] = `null`;

exports[`Then.jsx Then should render a single \`true\` child as \`null\` 2`] = `null`;

exports[`Then.jsx Then should render a single \`true\` child as \`null\` 3`] = `<div />`;

exports[`Then.jsx Then should render a single \`true\` child as \`null\` 4`] = `<div />`;

exports[`Then.jsx Then should render a single \`undefined\` child as \`null\` 1`] = `null`;

exports[`Then.jsx Then should render a single \`undefined\` child as \`null\` 2`] = `null`;

exports[`Then.jsx Then should render a single \`undefined\` child as \`null\` 3`] = `<div />`;

exports[`Then.jsx Then should render a single empty string child as "" 1`] = `""`;

exports[`Then.jsx Then should render a single empty string child as "" 2`] = `""`;

exports[`Then.jsx Then should render a single empty string child as "" 3`] = `
<div>
  
</div>
`;

exports[`Then.jsx Then should render a single non-empty string child as its string value 1`] = `"0"`;

exports[`Then.jsx Then should render a single non-empty string child as its string value 2`] = `"0"`;

exports[`Then.jsx Then should render a single non-empty string child as its string value 3`] = `"non-empty"`;

exports[`Then.jsx Then should render a single non-empty string child as its string value 4`] = `"non-empty"`;

exports[`Then.jsx Then should render a single non-empty string child as its string value 5`] = `"non-empty"`;

exports[`Then.jsx Then should render a single non-empty string child as its string value 6`] = `"non-empty"`;

exports[`Then.jsx Then should render a single non-zero number child as its string value 1`] = `"-Infinity"`;

exports[`Then.jsx Then should render a single non-zero number child as its string value 2`] = `"-1.7976931348623157e+308"`;

exports[`Then.jsx Then should render a single non-zero number child as its string value 3`] = `"-9007199254740991"`;

exports[`Then.jsx Then should render a single non-zero number child as its string value 4`] = `"-42"`;

exports[`Then.jsx Then should render a single non-zero number child as its string value 5`] = `"-1"`;

exports[`Then.jsx Then should render a single non-zero number child as its string value 6`] = `"1"`;

exports[`Then.jsx Then should render a single non-zero number child as its string value 7`] = `"42"`;

exports[`Then.jsx Then should render a single non-zero number child as its string value 8`] = `"9007199254740991"`;

exports[`Then.jsx Then should render a single non-zero number child as its string value 9`] = `"1.7976931348623157e+308"`;

exports[`Then.jsx Then should render a single non-zero number child as its string value 10`] = `"Infinity"`;

exports[`Then.jsx Then should render a single non-zero number child as its string value 11`] = `
<div>
  42
</div>
`;

exports[`Then.jsx Then should render a single zero child as \`0\` 1`] = `"0"`;

exports[`Then.jsx Then should render a single zero child as \`0\` 2`] = `"0"`;

exports[`Then.jsx Then should render a single zero child as \`0\` 3`] = `
<div>
  0
</div>
`;

exports[`Then.jsx Then should render an array child as expected based on type 1`] = `
Array [
  "Hello!",
  " ",
  "Today I am ",
  "42",
  " years old!",
]
`;

exports[`Then.jsx Then should render an array child as expected based on type 2`] = `
<div>
  Hello!
   
  Today I am 
  42
   years old!
</div>
`;

exports[`Then.jsx Then should render an empty array child as if there were no children 1`] = `null`;

exports[`Then.jsx Then should render an empty array child as if there were no children 2`] = `<div />`;

exports[`Then.jsx Then should render with no props and no children 1`] = `null`;

exports[`Then.jsx Then should throw with a single \`Error\` child 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#C00",
      "boxSizing": "border-box",
      "color": "#FFF",
      "cursor": "help",
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "maxHeight": "100vh",
      "maxWidth": "100vw",
      "textAlign": "center",
      "width": "100%",
    }
  }
  title="Invariant Violation: Objects are not valid as a React child (found: Error: Oh oh!). If you meant to render a collection of children, use an array instead.
    in Then
    in ErrorBoundary

This is located at:
    in Then
    in ErrorBoundary"
>
  <svg
    preserveAspectRatio="xMidYMid"
    style={
      Object {
        "fill": "currentColor",
        "flex": "1 1 auto",
      }
    }
    viewBox="0 0 24 24"
  >
    <path
      d="M20,12A8,8 0 0,0 12,4A8,8 0 0,0 4,12A8,8 0 0,0 12,20A8,8 0 0,0 20,
        12M22,12A10,10 0 0,1 12,22A10,10 0 0,1 2,12A10,10 0 0,1 12,2A10,10 0 0,1 22,
        12M15.5,8C16.3,8 17,8.7 17,9.5C17,10.3 16.3,11 15.5,11C14.7,11 14,10.3 14,
        9.5C14,8.7 14.7,8 15.5,8M10,9.5C10,10.3 9.3,11 8.5,11C7.7,11 7,10.3 7,9.5C7,
        8.7 7.7,8 8.5,8C9.3,8 10,8.7 10,9.5M12,14C13.75,14 15.29,14.72 16.19,
        15.81L14.77,17.23C14.32,16.5 13.25,16 12,16C10.75,16 9.68,16.5 9.23,
        17.23L7.81,15.81C8.71,14.72 10.25,14 12,14Z"
    />
  </svg>
</div>
`;
